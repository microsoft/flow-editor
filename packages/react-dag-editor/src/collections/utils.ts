export const is = Object.is;

export class MapIterator<T, U> implements IterableIterator<U> {
  private readonly upstream: Iterator<T>;
  private readonly f: (value: T) => U;

  public constructor(upstream: Iterator<T>, f: (value: T) => U) {
    this.upstream = upstream;
    this.f = f;
  }

  public [Symbol.iterator](): IterableIterator<U> {
    return this;
  }

  public next(): IteratorResult<U> {
    const next = this.upstream.next();
    if (next.done) {
      return next;
    }
    return {
      done: false,
      value: this.f(next.value)
    };
  }
}
