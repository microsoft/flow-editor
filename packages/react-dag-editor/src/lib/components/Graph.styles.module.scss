// use static css file here
// if we use css-in-js, classNames are generated in Graph component
// while node dragged out from module list can't get then className

.svg {
  position: absolute;
  overflow: visible;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;

  &:focus {
    outline: none;
  }
}

.node {
  cursor: move;
}

.container {
  position: relative;
  width: 100%;
  height: 100%;
  overflow: hidden;
  touch-action: none;
}

.button-a11y {
  opacity: 0;
  width: 0;
  height: 0;
  overflow: hidden;
}

.adding-node-svg {
  z-index: 1000000;
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
}

.module-item {
  user-select: none;
  cursor: pointer;
}

.minimap {
  height: 320px;
  width: 320px;
  user-select: none;
  touch-action: none;

  &-svg {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }
}
