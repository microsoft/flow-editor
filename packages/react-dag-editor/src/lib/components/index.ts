export * from "./ReactDagEditor";
export * from "./Graph/Graph";
export * from "./ItemPanel";
export * from "./Minimap/Minimap";
export * from "./StaticGraph/StaticGraph";
export * from "./Graph/IGraphProps";
export * from "./RegisterComponent";
export { defaultGetPositionFromEvent } from "../controllers";
export * from "./GridBackground";
export * from "./NodeAnchors";
