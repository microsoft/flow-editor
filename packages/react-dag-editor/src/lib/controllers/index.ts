export * from "./DragController";
export * from "./DragNodeController";
export * from "./TouchController";
export * from "./TwoFingerHandler";
