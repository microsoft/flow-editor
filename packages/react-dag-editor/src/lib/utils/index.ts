export * from "./getPortPosition";
export * from "./getNodeConfig";
export * from "./graphDataUtils";
export * from "./transformMatrix";
export * from "./getCurvePathD";
export * from "./layout";
export * from "./browsers";
export * from "./isWithinThreshold";
export * from "./geometric";
export * from "./viewport";
export * from "./Writable";
export * from "./gantt/getCriticalPath";
export * from "./gantt/getTopoSortingNodes";
export * from "./getOffsetLimit";
export * from "./event";
export * from "./valueOrCallback";
export * from "./debug";
export * from "./debounce";
export * from "./arrayEquals";
export * from "./empty";
export * from "./stopPropagation";
export * from "./clamp";
export * from "./history";
export * as Bitset from "./bitset";
