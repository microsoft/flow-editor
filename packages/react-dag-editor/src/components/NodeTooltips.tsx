import * as React from "react";
import { useGraphConfig } from "../hooks/context";
import type { IViewport } from "../models/geometry";
import type { NodeModel } from "../models/NodeModel";
import { GraphNodeStatus } from "../models/status";
import { getNodeConfig } from "../utils";
import * as Bitset from "../utils/bitset";

interface IProps {
  node?: NodeModel;
  viewport: Required<IViewport>;
}

export const NodeTooltips: React.FunctionComponent<IProps> = (props) => {
  const { node, viewport } = props;
  const graphConfig = useGraphConfig();

  if (!node) {
    return null;
  }

  if (!Bitset.has(GraphNodeStatus.Activated)(node.status)) {
    return null;
  }

  const nodeConfig = getNodeConfig(node, graphConfig);

  if (!nodeConfig?.renderTooltips) {
    return null;
  }

  return (
    <div className="node-tooltips">
      {nodeConfig.renderTooltips({
        model: node,
        viewport,
      })}
    </div>
  );
};
