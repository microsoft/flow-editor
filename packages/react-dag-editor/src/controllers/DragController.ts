/* eslint-disable no-invalid-this */
import { IEventProvider, IGlobalMoveEventTypes } from "../event-provider/types";
import { IPoint } from "../models/geometry";
import { noop } from "../utils/noop";
import { animationFramed, IScheduledCallback } from "../utils/scheduling";

export type TGetPositionFromEvent = (e: MouseEvent) => IPoint;

export interface IOnDragMove<T = MouseEvent> {
  /**
   * clientX after getPositionFromEvent
   */
  clientX: number;
  /**
   * clientY after getPositionFromEvent
   */
  clientY: number;
  dx: number;
  dy: number;
  totalDX: number;
  totalDY: number;
  e: T;
}

export interface IOnDragEnd<T = MouseEvent> {
  totalDX: number;
  totalDY: number;
  e: T;
}

export class DragController<T extends MouseEvent = MouseEvent> {
  public onMove: (move: IOnDragMove) => void = noop;
  public onEnd: (end: IOnDragEnd) => void = noop;
  protected mouseMove: IScheduledCallback<[T]>;
  protected lastEvent: T | null = null;
  private readonly getPositionFromEvent: TGetPositionFromEvent;
  private startX = 0;
  private startY = 0;
  private prevClientX = 0;
  private prevClientY = 0;
  private readonly eventProvider: IEventProvider<IGlobalMoveEventTypes>;

  public constructor(
    eventProvider: IEventProvider<IGlobalMoveEventTypes>,
    getPositionFromEvent: TGetPositionFromEvent
  ) {
    this.eventProvider = eventProvider;
    this.getPositionFromEvent = getPositionFromEvent;
    this.mouseMove = animationFramed((e: T) => {
      this.doOnMouseMove(e);
    });
  }

  public start(e: T): void {
    this.lastEvent = e;
    const { x, y } = this.getPositionFromEvent(e);
    this.startX = x;
    this.startY = y;
    this.prevClientX = x;
    this.prevClientY = y;

    this.eventProvider.on("move", this.onMouseMove as (e: MouseEvent) => void);
    this.eventProvider.on("end", this.onMouseUp as (e: MouseEvent) => void);
  }

  public stop(): void {
    this.mouseMove.cancel();
    this.eventProvider.off("move", this.onMouseMove as (e: MouseEvent) => void);
    this.eventProvider.off("end", this.onMouseUp as (e: MouseEvent) => void);
  }

  protected getDelta(x: number, y: number): IPoint {
    const clientDX = x - this.prevClientX;
    const clientDY = y - this.prevClientY;
    this.prevClientX = x;
    this.prevClientY = y;
    return {
      x: clientDX,
      y: clientDY,
    };
  }

  protected getTotalDelta(e: MouseEvent): IPoint {
    const x = e.clientX - this.startX;
    const y = e.clientY - this.startY;
    return {
      x,
      y,
    };
  }

  protected doOnMouseMove(e: MouseEvent): void {
    const { x, y } = this.getPositionFromEvent(e);
    const { x: dx, y: dy } = this.getDelta(x, y);
    const { x: totalDX, y: totalDY } = this.getTotalDelta(e);
    this.onMove({
      clientX: x,
      clientY: y,
      dx,
      dy,
      totalDX,
      totalDY,
      e,
    });
  }

  protected doOnMouseUp(e: T): void {
    e.preventDefault();
    const { x: totalDX, y: totalDY } = this.getTotalDelta(e);
    this.onEnd({ totalDX, totalDY, e });

    this.stop();
  }

  private readonly onMouseUp = (e: T) => {
    this.lastEvent = e;
    this.doOnMouseUp(e);
    this.lastEvent = null;
  };

  private readonly onMouseMove = (e: T) => {
    this.lastEvent = e;
    e.preventDefault();
    this.mouseMove(e);
  };
}

export function defaultGetPositionFromEvent<T extends MouseEvent>(
  e: T
): IPoint {
  return {
    x: e.clientX,
    y: e.clientY,
  };
}
