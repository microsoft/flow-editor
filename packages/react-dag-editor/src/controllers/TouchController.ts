/* eslint-disable no-invalid-this */
import * as React from "react";
import { animationFramed } from "../utils/scheduling";

export interface ITouchHandler<ExtraArgs extends unknown[] = []> {
  onStart(
    events: Map<number, PointerEvent>,
    e: PointerEvent,
    ...args: ExtraArgs
  ): void;
  onMove(
    events: Map<number, PointerEvent>,
    e: PointerEvent,
    ...args: ExtraArgs
  ): void;
  onEnd(e: PointerEvent, ...args: ExtraArgs): void;
}

export class TouchController<ExtraArgs extends unknown[] = []> {
  public readonly eventHandlers = {
    onPointerDown: (e: React.PointerEvent, ...args: ExtraArgs): void => {
      if (e.pointerType !== "touch") {
        return;
      }
      e.preventDefault();
      this.pointers = new Map(this.pointers);
      this.pointers.set(e.pointerId, e.nativeEvent);
      this.updateHandler(e.nativeEvent, ...args);
    },

    onPointerMove: (e: React.PointerEvent, ...args: ExtraArgs): void => {
      if (e.pointerType !== "touch") {
        return;
      }
      e.preventDefault();
      this.pointers.set(e.pointerId, e.nativeEvent);
      this.onMove(e.nativeEvent, ...args);
    },

    onPointerUp: (e: React.PointerEvent, ...args: ExtraArgs): void => {
      if (e.pointerType !== "touch") {
        return;
      }
      e.preventDefault();
      this.pointers = new Map(this.pointers);
      this.pointers.delete(e.pointerId);
      this.updateHandler(e.nativeEvent, ...args);
    }
  };

  private pointers = new Map<number, PointerEvent>();
  private currentHandler: ITouchHandler<ExtraArgs> | undefined;
  private readonly handlers: Map<number, ITouchHandler<ExtraArgs>>;

  private readonly onMove = animationFramed(
    (e: PointerEvent, ...args: ExtraArgs) => {
      this.currentHandler?.onMove(this.pointers, e, ...args);
    }
  );

  public constructor(handlers: Map<number, ITouchHandler<ExtraArgs>>) {
    this.handlers = handlers;
  }

  private updateHandler(e: PointerEvent, ...args: ExtraArgs): void {
    const handler = this.handlers.get(this.pointers.size);
    if (handler !== this.currentHandler) {
      this.currentHandler?.onEnd(e, ...args);
      this.currentHandler = handler;
      this.currentHandler?.onStart(this.pointers, e, ...args);
    }
  }
}
