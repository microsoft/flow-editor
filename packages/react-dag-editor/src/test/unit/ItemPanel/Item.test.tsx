/* eslint-disable @typescript-eslint/no-explicit-any */
import {
  act,
  cleanup,
  fireEvent,
  render,
  RenderResult,
  screen,
} from "@testing-library/react";
import * as React from "react";
import {
  GraphCanvasEvent,
  GraphConfigBuilder,
  GraphModel,
  rect,
} from "../../../index";
import { Item } from "../../../src/components/ItemPanel";
import { GraphController } from "../../../src/controllers/GraphController";
import { GraphControllerRef, TestComponent } from "../../TestComponent";
import { mockClientRect, patchPointerEvent } from "../../utils";
import { TestItemContent } from "./TestItemContent";

jest.mock("../../../src/components/ItemPanel/useSvgRect", () => ({
  useSvgRect: () => {
    return mockClientRect;
  },
}));

describe("ItemPanel - Item", () => {
  let dragWillStart: () => void;
  let renderedWrapper: RenderResult;
  let graphController: GraphController;
  const getElement = () => renderedWrapper.getByRole("button");

  beforeAll(() => {
    jest.useFakeTimers();
    patchPointerEvent();
  });
  afterAll(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    dragWillStart = jest.fn();
    const graphConfig = GraphConfigBuilder.default()
      .registerNode("nodeShape", rect)
      .build();
    const graphControllerRef = React.createRef<GraphController>();
    renderedWrapper = render(
      <TestComponent
        data={GraphModel.empty()}
        graph={false}
        settings={{ graphConfig }}
      >
        <Item
          model={{ name: "node1", shape: "nodeShape" }}
          dragWillStart={dragWillStart}
          nodeWillAdd={jest.fn()}
          nodeDidAdd={jest.fn()}
        >
          <TestItemContent text="test item" />
        </Item>
        <GraphControllerRef ref={graphControllerRef} />
      </TestComponent>
    );
    graphController = graphControllerRef.current!;
    expect(graphController).toBeDefined();
    act(() => {
      graphController.dispatch({
        type: GraphCanvasEvent.ViewportResize,
        viewportRect: mockClientRect,
      });
    });
    (graphController.eventChannel.listenersRef as any).current =
      graphController.dispatch;
  });

  afterEach(cleanup);

  it("Should work well when mousedown and mouseup", () => {
    expect(getElement()).toMatchSnapshot();

    // when mouse down
    act(() => {
      fireEvent.pointerDown(screen.getByRole("button"));
    });
    expect(dragWillStart).toBeCalled();
    expect(getElement()).toMatchSnapshot();

    // when mouse up
    act(() => {
      jest.runAllTimers();
      fireEvent.pointerUp(window);
    });
    expect(getElement()).toMatchSnapshot();
  });
});
