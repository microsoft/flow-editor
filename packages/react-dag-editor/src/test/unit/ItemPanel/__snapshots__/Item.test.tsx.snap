// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemPanel - Item Should work well when mousedown and mouseup 1`] = `
<div
  class="moduleItem-5"
  role="button"
>
  <p>
    test item:default
  </p>
</div>
`;

exports[`ItemPanel - Item Should work well when mousedown and mouseup 2`] = `
<div
  class="moduleItem-5"
  role="button"
>
  <p>
    test item:addingNode
  </p>
</div>
`;

exports[`ItemPanel - Item Should work well when mousedown and mouseup 3`] = `
<div
  class="moduleItem-5"
  role="button"
>
  <p>
    test item:default
  </p>
</div>
`;

exports[`ItemPanel - Item Should work well when mousedown and mouseup: events 1`] = `[]`;
